src/main/kotlin/racingcar/domain/Analyzer.kt

```diff
@@ -0,0 +1,28 @@
+package racingcar.domain
+
+import racingcar.data.Car
+
+class Analyzer {
+    fun judgeWinner(cars: List<Car>): List<Car> {
+        val winners = mutableListOf<Car>()
+        var maxLocation = NOT_YET_SET
+
+        for (car in cars) {
+            if (car.location > maxLocation) {
+                winners.clear()
+                winners.add(car)
+                maxLocation = car.location
+            } else if (car.location == maxLocation) {
+                winners.add(car)
+            }
+        }
+        return winners.toList()
+    }
+
+    fun isMoveAllowed(num: Int): Boolean = num >= MIN_PASSING_NUM
```

Comment:
[3] [컨벤션](https://kotlinlang.org/docs/coding-conventions.html#avoid-redundant-constructs)에서는 `"명확성을 위해" 코드에 불필요한 구문 요소를 남겨두지 마십시오.` 라는 지침이 있어 반환 타입을 제거하는 게 어떨까 합니다. 
물론 가독성을 고려한다면 반환 타입을 넣는 게 맞지 않을까? 라는 고민을 했었지만, 표현식을 보면 1초도 안돼서 직관적으로 Boolean 타입이라는 걸 알 수 있기도 해서 저는 반환 타입을 제거하고 있습니다.
```suggestion
    fun isMoveAllowed(num: Int) = num >= MIN_PASSING_NUM
```

Reply:
리턴 값을 명시하지 않았을 때, IDE 에러가 나와서 명시하도록 수정한 적이 있는데, 잘 되네요..😅

Reply:
이번 4주차 미션을 진행하면서, 리턴 타입을 명시하지 않았을 때, Junit5 테스트 코드에서 Unresolved reference 에러가 발생하는 경험을 했습니다. Application.kt으로 실행할 때는 에러가 발생하지 않았습니다.

 구글에 검색해도 비슷한 사례나 해결책을 찾지 못했습니다. 저의 추측으로는 리턴 타입이 한 개로 분명하지만, null을 리턴하는 고차 함수를 사용하면 컴파일러의 타입 추론에 문제가 생기는 것 같아요.

```kotlin
package christmas.data

enum class Badge(val nickName: String, val minBenefitAmount: Int) {
    NONE("없음", 0), STAR("별", 5_000),
    TREE("트리", 10_000), SANTA("산타", 20_000);

    companion object {
        fun from(benefitAmount: Int): Badge = entries.sortedBy { // 여기서  Badge 리턴 타입을 명시하지 않으면 테스트 코드에서 에러 발생
            -it.minBenefitAmount
        }.firstOrNull {
            it.minBenefitAmount <= benefitAmount
        } ?: NONE
    }
}

class BadgeTest {
    
    @ParameterizedTest
    @ValueSource(ints = [20_000, 30_000, 999_999])
    @DisplayName("Badge - SANTA")
    fun `2만 원 이상의 혜택을 받으면 SANTA를 반환한다`(benefitAmount: Int) {
        // when
        val actual: Badge = Badge.from(benefitAmount) // Unresolved reference 에러가 발생했던 곳

        // then
        val expected = Badge.SANTA
        Assertions.assertThat(actual).isEqualTo(expected)
    }

```